### ISSUE TITLE
Error: self signed certificate in certificate chain

### ISSUE THREAD

**Issue by codechaotic** [apocas/dockerode#154]
While trying to use dockerode with Joyent's Triton Elastic Container Service I have setup the following really basic script to test my configurations:

```
//app.js
var Docker = require('dockerode')

var docker = new Docker()

docker.info( function(err, info) {
  console.log(err)
  if(err) throw err
  console.log(info)
});
```

The configurations themselves are setup using a helper script provided in the documentation for Joyent's container service [here](https://apidocs.joyent.com/docker). So the above script is pulling configuration from the environment variables `DOCKER_CERT_PATH`, `DOCKER_HOST`, and `DOCKER_TLS_VERIFY` which the helper script sets up after creating all the necessary certificates. (I should also note that I tried setting up the configurations explicitly with no change in behaviour)

When the script runs it produces the following output:

```
{ [Error: self signed certificate in certificate chain] code: 'SELF_SIGNED_CERT_IN_CHAIN' }
/path/to/app.js:32
  if(err) throw err
                ^
Error: self signed certificate in certificate chain
    at Error (native)
    at TLSSocket.<anonymous> (_tls_wrap.js:929:36)
    at TLSSocket.emit (events.js:104:17)
    at TLSSocket._finishInit (_tls_wrap.js:460:8)
```

It should be noted that the configurations produced by the helper script are working when I use docker directly. For example running `docker info` prints out

```
$> docker info
Containers: 0
Images: 0
Storage Driver: sdc
 SDCAccount: #########
Execution Driver: sdc-0.3.0
Kernel Version: 3.12.0-1-amd64
Operating System: SmartDataCenter
CPUs: 0
Total Memory: 0 B
Name: us-east-1 
```

which properly represents the remote host.

A few details about my environment:
- I have updated Node and NPM to the newest versions (v0.12.6 and 2.12.1)
- I am running these tests on Linux Mint 17.1
- I am using Docker version 1.7.0, build 0baf609

I'm really not very savvy with SSL certificates, and I've had very little luck working out the issue on my own. I would appreciate any tips or ideas that could help work past this error. Thanks!

**Comment by codechaotic** [comment]
Listed `DOCKER_CERT_PATH` twice

**Comment by codechaotic** [comment]
I've got an update and solution!

I'm sure there is a more elegant solution than this, but adding the following line at the top of my code fixes the error.
```
process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0"
```
After this I received an http error that the endpoint wasn't found:
```
/path/to/app.js:19
  if(err) throw err
                ^
Error: HTTP code is 404 which indicates error: undefined
 - (ResourceNotFound) /info does not exist (563ad819-2393-4777-836a-dbd17b582ea2)
...stack trace...
```
A little more experimentation and I found that `curl`ing the Docker Remote API directly doesn't work either UNLESS the endpoint explicitly declares the version. 

So this works:
```
curl https://us-east-1.docker.joyent.com:2376/v1.19/info --cert cert.pem \
       --key key.pem --cacert ca.pem
```
But this does not:
```
curl https://us-east-1.docker.joyent.com:2376/info  --cert cert.pem \
       --key key.pem --cacert ca.pem
```
To make sense of this it took realizing that Joyent isn't running vanilla docker at all, but rather a reimplementation which better suits their services (which they call their SmartDataCenter) called [sdc-docker](https://github.com/joyent/sdc-docker). Although they have done an amazing job of making their service 'Bug for Bug compatible' with the Docker Remote API there are known divergences. Most of these are well [documented](https://github.com/joyent/sdc-docker/blob/master/docs/api/divergence.md#divergence).

While I didn't find any mention of the version being required when sending requests to Joyent's version of the Remote API, this seems to be another small divergence that escaped mention. To this effect, the following rewrite of my original program works like a charm:
```
process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";

var Docker = require('dockerode'),
    fs = require('fs')

var docker = new Docker({
  protocol: 'https',
  version: 'v1.19',
  port: 2376,
  host: 'us-east-1.docker.joyent.com',
  ca: fs.readFileSync('ca.pem'),
  cert: fs.readFileSync('cert.pem'),
  key: fs.readFileSync('key.pem')
})

docker.info( function(err, info) {
  if(err) throw err
  console.log(info)
})

**Comment by apocas** [comment]
Wow I wasn't aware this kind of fragmentation existed in here.

Not much we can do in dockerode I guess, since we must follow original Docker spec.

Awesome investigation BTW! I may be very useful to someone that has the same use case.

**Comment by codechaotic** [comment]
Thanks, and I'm hoping so. Joyent is still converging their version of the Docker API. The above related issue was closed because the required version issue had already been noticed and is fixed for a future release. See https://smartos.org/bugview/DOCKER-441

What really would have saved me some time in my investigations is better documentation of the accepted parameters of the docker-modem object. I had actually created a fork of dockerode just to force requests to specify a version in the path. Only after that, by digging through the docker-modem sourcecode to understand how the package works did I realize that I could just specify a version parameter. The examples in the documentation for dockerode are great, but listing ALL allowed options would be helpful.

Heck I'll make a PR if you're cool with it.
